### DISCUSSION TITLE
icns File Upload?

### DISCUSSION THREAD

**Discussion by rodgerwerner** [elrumo/macOS_Big_Sur_icons_replacements#1098] [Ideas]
Would it be possible to allow icns file upload? The main reason I ask is because sometimes smaller sized icons have slightly different designs. One example of this is the Logitech Options icon where the main larger icon shows multiple devices and the smaller 32px & 16px designs show just a mouse.

**Comment by elrumo** [comment]
I've thought of that before, and it's something that would be simple to implement. The only issue that I have is that I could not find a way to create a png image to display back on the website from an .icns file. At the moment I'm using a node library to convert .png to .icns, but I couldn't find one that did the opposite, so if you or anyone knows of a library that can do just that let me know.

**Reply by setanarut** [reply]
This Golang package can encode and decode ICNS https://github.com/JackMordaunt/icns#command-line

**Reply by joroldan** [reply]
Native macOS command "iconutil" can convert ICNS to PNG. It creates a folder with all png files included in the icon. This is achieved using parameter "-c iconset".

**Reply by bunkerdesign** [reply]
I used to upload .icns files with no problems before the update

**Reply by elrumo** [reply]
.icons have never been supported, at least intentioanlly, so I'd be very curious to see what was it that worked. Do you have an example of an icon you submited? Also, what is your username so I could look it up

**Reply by bunkerdesign** [reply]
> .icons have never been supported, at least intentioanlly, so I'd be very curious to see what was it that worked. Do you have an example of an icon you submited? Also, what is your username so I could look it up

https://macosicons.com/#/u/pipofezine
all icons before Marmoset Toolbag 4

**Comment by setanarut** [comment]
related Issue https://github.com/elrumo/macOS_Big_Sur_icons_replacements/issues/1084#issue-803132333

**Comment by joaofrgomes** [comment]
Please implement a parser which actually respects and directly uploads properly formatted .icns files without any sort of processing… I've just downloaded a file I've submitted, and dug back into its instances to compare them to the original ones, and the 16x16@1x instance in particular is mangled beyond recognition (at least when you zoom in, or see it on a non-Retina screen).

Here, take a look; the original:
![icon_16x16](https://user-images.githubusercontent.com/29361091/154972002-c5e1ce14-f5d7-4c5a-96dd-3dee14948b27.png)

The same instance, as processed by macOSicons:
![icon_16x16](https://user-images.githubusercontent.com/29361091/154972099-68a19c8a-b7b1-4d7d-88ef-7e8410b6b82e.png)

And this isn't even the worst offender… I have several icons which are actually even more hinted for non-Retina screens (as those will be a thing for quite some time, still).

There are other details, such as resolution-specific textures that are also not being respected either, of course…

So, it seems your website is only taking the PNG from the 512x512@2x instance and downscaling it, is my assumption correct? That is an issue, because if you look at Apple's template, that's not actually how icons are supposed to be drawn; besides the aforementioned hinting, the transparent border doesn't scale linearly, either.

Also, regarding icons already in the database, will we be able to update their files to our original ones, to at least keep the download count? Perhaps with a little “updated” badge or something, a reason for editing and maybe even a section/tag for that, so people can actually choose to re-download them?

**Comment by elrumo** [comment]
Apologies for the delayed response, I’m using a node library to create the .icns file. I’ll have another look at the options it has in case I missed some something to select the right colour space. I’ll keep you updated, but if you come across any other node library that can handle .icns better, let me know. 

Also, the library that I’m currently using only goes one way: .png to .icns, which is the reason I haven’t enabled .icns uploads yet as I’d need a way to generate the low res .png preview. Another option could be to let users upload the png preview themselves as well as the .icns, but that opens the door for a multitude of othe user-generated issues. I’ll have a think.

**Reply by joaofrgomes** [reply]
Well, your library can already convert .icns to .png and back to .icns just fine; you will just have to code your parser to open an exception and leave .icns uploads well alone in the final product, while using the output of your current system to generate just the previews. Simple enough?

Maybe it does mean customising the library, and maybe that's a bit too much, but the whole conversion part is already feasible because it's already happening behind the scenes without any input on our part. You really don't need to complicate things further by allowing users to input .png thumbnail files by hand.

**Reply by isametry** [reply]
> Another option could be to let users upload the png preview themselves as well as the .icns, but that opens the door for a multitude of othe user-generated issues.

Maybe you could enable this as a closed beta for now? Only allow a couple of creators to opt-in, and see how it goes. 

And if you don't mind me asking, what would you say are the potential problems connected to this? The resolution can be checked before upload, and compression + color conversion can be done on the server side. Are you thinking of inaccurate, or even intentionally misleading thumbnails? Wouldn't the approval process prevent this?

**Reply by elrumo** [reply]
It took a while, but I finally made the time and the page now let's you upload .icns files directly. Hope it's what you guys expected.

**Reply by joaofrgomes** [reply]
Hi again! ~~Well, I would love to try it out, but I can't seem to find it. Maybe you just don't have a visible link to it?~~
[Edit: nevermind, I've already found it. Thank you for your support, good job!]

Also, I would very much like to be able to replace the files converted from .icns to .png and back to .icns with my original files in my already submitted entries, so that I can maintain the download count on them. Do you figure that's also possible? I did try selecting my .icns files again, but it seems they are still not recognized as supported in the file browser and were, thus, grayed-out and unclickable.

<img width="1436" alt="Captura de ecrã 2022-07-28, às 03 46 28" src="https://user-images.githubusercontent.com/29361091/181409223-b69f2bae-6790-4c4f-80bd-b97183e562e1.png">

**Reply by elrumo** [reply]
.icns upload only works for new icons right now, but I will eventually add it to replace existing icons too when I get more time. I'll let you know once it's up.

**Reply by joaofrgomes** [reply]
Hi! I've just detected another bug related to this issue, so I figured that it wouldn't be worth cluttering the discussions with a new one.

When trying to upload multiple files in a single pass, the .icns upload feature broke the whole thing.

If you upload .icns files, I believe the last file to be uploaded will be assigned to a wrong entry (the first one?) and the others will become “phantom” entries with no files associated to them (either converted .png-to-.icns or original .icns files; when clicking entries, instead of downloading anything you're just taken to the main page).
